pkg/migration-tool/reader/reader.go

```diff
@@ -0,0 +1,174 @@
+		Timeout: model.Duration(DefaultReadTimeout),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("creating remote-client: %w", err)
+	}
+	read := &RemoteRead{
+		url:           readStorageUrl,
+		plan:          p,
+		client:        rc,
+		sigBlockRead:  sigRead,
+		sigBlockWrite: sigWrite,
+	}
+	read.fetchersRunning.Store(0)
+	return read, nil
+}
+
+// Run runs the remote read and starts fetching the samples from the read storage.
+func (rr *RemoteRead) Run(wg *sync.WaitGroup, readerUp *atomic.Bool) error {
+	defer func() {
+		close(rr.sigBlockRead)
+		close(rr.sigBlockWrite)
+		readerUp.Store(false)
+		wg.Done()
+	}()
+	readerUp.Store(true)
+	log.Info("msg", "reader is up")
+	timeRangeMinutesDelta := time.Minute.Milliseconds()
+	for i := rr.plan.Mint; i <= rr.plan.Maxt; {
+		var (
+			mint     = i
+			maxt     = mint + timeRangeMinutesDelta
+			fetch    = rr.createFetch(rr.url, mint, maxt)
+			blockRef = rr.plan.CreateBlock(mint, maxt)
+		)
+		blockRef.SetDescription(fmt.Sprintf("fetching time-range: %d mins...", timeRangeMinutesDelta/time.Minute.Milliseconds()), 1)
+		result, err := fetch.start(context.Background(), []*labels.Matcher{labels.MustNewMatcher(labels.MatchRegexp, "", ".*")})
+		if err != nil {
+			return fmt.Errorf("remote-read run: %w", err)
+		}
+		if len(result.Timeseries) == 0 {
+			rr.plan.DecrementBlockCount()
+			i += timeRangeMinutesDelta + 1
+			if timeRangeMinutesDelta < MaxTimeRangeDeltaLimit.Milliseconds() {
+				timeRangeMinutesDelta *= 2
+			}
+			continue
+		}
+		qResultSize := result.Size() // Assuming each character corresponds to one byte.
+		rr.plan.UpdateLastPushedMaxt(i)
+		blockRef.SetDescription(fmt.Sprintf("received %d MB with delta %d mins...", qResultSize/megaByte, timeRangeMinutesDelta/time.Minute.Milliseconds()), 1)
+		blockRef.Timeseries = result.Timeseries
+		rr.sigBlockRead <- struct{}{}
+		<-rr.sigBlockWrite
+		if err := rr.plan.Clear(); err != nil {
+			return fmt.Errorf("remote-read run: %w", err)
+		}
+		switch {
+		case qResultSize < ResponseDataSizeHalfLimit && timeRangeMinutesDelta >= MaxTimeRangeDeltaLimit.Milliseconds():
+			i += timeRangeMinutesDelta + 1
+		case qResultSize < ResponseDataSizeHalfLimit:
+			i += timeRangeMinutesDelta + 1
+			timeRangeMinutesDelta *= 2
+		default:
+			i += timeRangeMinutesDelta + 1
+			timeRangeMinutesDelta /= 2
+		}
+	}
+	log.Info("msg", "reader is down")
+	return nil
+}
+
+type fetch struct {
+	url        string
+	mint, maxt int64
+	clientCopy remote.ReadClient // Maintain a copy of remote client for parallel fetching.
+}
+
+func (rr *RemoteRead) createFetch(url string, mint, maxt int64) *fetch {
+	f := &fetch{
+		url:        url,
+		mint:       mint,
+		maxt:       maxt,
+		clientCopy: rr.client,
+	}
+	rr.fetchersRunning.Add(1)
+	return f
+}
+
+// start starts fetching the samples from remote read storage as client based on the matchers.
+func (f *fetch) start(context context.Context, matchers []*labels.Matcher) (*prompb.QueryResult, error) {
```

Comment:
looks like the function starts and finishes fetching the samples, maybe it can be called fetch